pkg/storage/rulestore/config.go

```diff
@@ -0,0 +1,122 @@
+package rulestore
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/thanos/pkg/objstore"
+
+	"github.com/cortexproject/cortex/pkg/configs/client"
+	"github.com/cortexproject/cortex/pkg/ruler/rules"
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/azure"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/filesystem"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/gcs"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/s3"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/swift"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const (
+	ConfigDB = "configdb"
+
+	name   = "rule-store"
+	prefix = "rule-store."
+)
+
+var (
+	supportedBackends = []string{bucket.S3, bucket.GCS, bucket.Azure, bucket.Swift, bucket.Filesystem, ConfigDB}
+
+	ErrUnsupportedStorageBackend = errors.New("unsupported storage backend")
+)
+
+// Config configures a rule store.
+type Config struct {
+	bucket.Config `yaml:",inline"`
+	ConfigDB      client.Config `yaml:"configdb"`
+}
+
+// RegisterFlags registers the backend storage config.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.ConfigDB.RegisterFlagsWithPrefix(prefix, f)
+
+	cfg.S3.RegisterFlagsWithPrefix(prefix, f)
```

Comment:
You should call `cfg.Bucket.RegisterFlagsWithPrefix()` instead of duplicating here.

Reply:
The only reason I didn't call this function is because it wouldn't print `configdb` as an option in the help text.

Between this issue and https://github.com/cortexproject/cortex/pull/3805#discussion_r574604170 I think I found a solution that will accommodate the use case case.

1. Add `ExtraBackends []string` field to the bucket.Config struct.
2. In `rulestore.Config.RegisterFlags` we set the `bucket.Config.ExtraBackends` field to `[]string{ConfigDB}`
3. The `bucket.Config.RegisterFlagsWithPrefix` and `bucket.Config.Validate` functions are adjusted to merge the `supportedBackends` with the embedded `ExtraBackends` before validating/printing-help-text

Let me know what you think? I think this approach will avoid any additional functions at the cost of an additional hidden field in the config that must be set before the `bucket.Config.Validate` or `bucket.Config.RegisterFlags` functions are called.